### PR TITLE
Bill Generation

### PR DESCRIPTION
**Pull request by aivantg**
[//]: # "These comments meant for your reference, they are invisible and don't need to be deleted"

## What's new in this PR?

This PR completes the backend aspect of bill generation (excluding PDFs)! This process looks like this: 

1. A post request is sent to the `/generate` route with one body parameter: `solarprojectId`
2. Backend pulls all subscribers associated with that solar project. For each susbcriber it:
3. Pulls their latest PG&E bill (and verifies that it hasn't processed this bill before)
4. Pulls the subscriber's enphase production data for the date range of the bill
5. Creates a Subscriber Bill Record using data from PG&E, Enphase, and the Previous Subscriber Bill

It logs its actions along the way so that we can come back to it easily!

Also installed the Airtable Schema Generator Package

[//]: # "Describe what's new in this PR"

### How to Review

Just logically follow those steps! Keep an eye out for the frontend PR that will help trigger this process.

### Related PRs

[#87](https://github.com/calblueprint/peoplepower-web/pull/87) - Frontend PR!
### Next steps

Need to add PDF Generation

### Tests Performed, Edge Cases

I looked at the data and made sure it was computing the right numbers and that it used the old bill's numbers correctly!

CC: @dfangshuo

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,24 +1,30 @@
...
-const fetch = require('node-fetch'); // Import fetch (for network requests)
-const moment = require('moment');
+import dotenv from 'dotenv-safe';
+import express from 'express';
+import cors from 'cors';
+import generateBillsForSolarProject from './utils/billgeneration';
 
-const EnphaseUtils = require('./utils/enphase.js');
-const UtilityAPI = require('./utils/utility-api.js');
+dotenv.config(); // Set up environment variables
```
Comment (dfangshuo) [inline index.js]: what environment variables are we using here?

File: utils/billgeneration.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  getOwnerById,
+  getRateScheduleById,
+  createSubscriberBill,
+  updateSubscriberBill,
+  getSubscriberBillsByStatementNumber
+} from '../airtable/request';
+import getEnphaseData from './enphase';
+import getLatestBill from './utilityApi';
+
+dotenv.config();
```
Comment (dfangshuo) [inline utils/billgeneration.js]: similar qn as above

File: utils/billgeneration.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  const startDate = startMoment.format('MM/DD/YYYY');
+  const endDate = endMoment.format('MM/DD/YYYY');
+
+  let prevBill;
+  if (subscriber.latestBillNumber !== 0) {
+    const bills = await getSubscriberBillsByStatementNumber(
+      subscriber.latestBillNumber
+    );
+    [prevBill] = bills.filter(b => b.subscriberId[0] === subscriber.id);
```
Comment (dfangshuo) [inline utils/billgeneration.js]: im not familiar with this syntax - what does this do?

File: utils/billgeneration.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  // Set previous bill to previous
+  await updateSubscriberBill(prevBill.id, { status: 'Previous' });
+
+  console.log(
+    `Found PGE Data for ${subscriber.name}
+Net Usage: ${netPgeUsage}
+ebce Rebate: ${ebceRebate}
+start date: ${startDate}
+end date: ${endDate}`
```
Comment (dfangshuo) [inline utils/billgeneration.js]: nit: adjust the indentation for consistency?

File: utils/enphase.js [hunk 5]
```diff
@@ -4,9 +4,30 @@ const moment = require('moment');
 const baseUrl = 'https://api.enphaseenergy.com/api/v2';
 const key = process.env.ENPHASE_KEY;
 
+const EnphaseSettings = {
+  rechN8WXaM60Xx859: data => data.meter_production, // lora foo Unit B
+  recMzaIHIl5Uq1xRJ: data =>
+    data.micro_production.map((v, i) => v - data.meter_production[i]) // lora foo unit A
```
Comment (dfangshuo) [inline utils/enphase.js]: would this mean that when a new user is onboarded, we would need to update this object?

File: .prettierrc [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+  "singleQuote": true
```
[the remark this answers is not in the document]
Comment (aivantg) [reply .prettierrc]: I don’t actually remember making this decision, i think i inherited it from one of the tutorials i used on installing all of this? That said, i do kinda prefer single quotes and hate inconsistent quoting hehe so i think i chose to leave it in. 

I copied this file from peoplepower-web tho

File: index.js (hunk 1 above)
Comment (aivantg) [reply index.js]: I suppose none anymore, but the intention here was that we load environment variables into the app as soon as possible since this is the entry point of the app, as I think is best practice. 

I tried including it in JUST here and getting that to work for the rest of the app, but it wasn’t working. I’ll try again though.. ideally it’s only called in this file and no others

File: utils/billgeneration.js (hunk 3 above)
Comment (aivantg) [reply utils/billgeneration.js]: It is called “Array Destructuring!” Honestly, I’ve never understood the purpose for it, but this was one of the few cases where I saw it’s usefulness.

`const [a, b] = [1, 2, 3]`
Is equivalent to
```
const array = [1, 2, 3];
const a = array[0;
const b = array[1];
```

So in this case it just gets the first element from the list

File: utils/billgeneration.js (hunk 4 above)
Comment (aivantg) [reply utils/billgeneration.js]: This weird indentation is actually because with these special strings (and maybe for all strings?) the indentation gets included in the output, so those lines intentionally have no indentation for cleaner logs. Could alternatively format with “\n”s and “+”s but low priority i think

File: utils/enphase.js (hunk 5 above)
Comment (aivantg) [reply utils/enphase.js]: Yup. This was what was brought up in conversation with Grayson: until we can find a better way to query per-module, we’ll have to do it this way. Once we handle errors better, this will be an error that is brought up if you a subscriber doesn’t have something here
